src/libopensc/pkcs15-cac.c

```diff
@@ -0,0 +1,429 @@
+		       SC_PKCS15_PRKEY_USAGE_SIGN | SC_PKCS15_PRKEY_USAGE_NONREPUDIATION;
+#endif
+	}
+	return 0;
+}
+
+
+/* These are the cert key usage bits that map to various PKCS #11 (and thus PKCS #15) flags */
+#define SC_CERT_USAGE_SIGNATURE 0x63ULL /* Signature, Non repudication, Cert Sign, Crl sign */
+#define SC_CERT_USAGE_DERIVE    0x10ULL /* Key agreement */
+#define SC_CERT_USAGE_UNWRAP	0x14ULL /* Key encipherment, Key agreement */
+#define SC_CERT_USAGE_DECRYPT	0x88ULL /* Data encipherment, encipher */
+#define SC_CERT_USAGE_NONREPUDIATION 0x02ULL
+
+/* map a cert usage and algorithm to public and private key usages */
+static int
+cac_map_usage(unsigned long long cert_usage, int algorithm, unsigned int *pub_usage_ptr, unsigned int *pr_usage_ptr, int allow_nonrepudiation)
+{
+	unsigned int pub_usage = 0, pr_usage = 0;
+	unsigned int alg_flags = cac_alg_flags_from_algorithm(algorithm);
+
+	if (cert_usage & SC_CERT_USAGE_SIGNATURE) {
+		pub_usage |= SC_PKCS15_PRKEY_USAGE_VERIFY|SC_PKCS15_PRKEY_USAGE_VERIFYRECOVER;
+		pr_usage |= SC_PKCS15_PRKEY_USAGE_SIGN|SC_PKCS15_PRKEY_USAGE_SIGNRECOVER;
+	}
+	if (cert_usage & SC_CERT_USAGE_DERIVE) {
+		pub_usage |= SC_PKCS15_PRKEY_USAGE_DERIVE;
+		pr_usage |= SC_PKCS15_PRKEY_USAGE_DERIVE;
+	}
+	if (cert_usage & (SC_CERT_USAGE_DECRYPT|SC_CERT_USAGE_UNWRAP)) {
+		pub_usage |= SC_PKCS15_PRKEY_USAGE_ENCRYPT;
+		pr_usage |= SC_PKCS15_PRKEY_USAGE_DECRYPT;
+	}
+	if (allow_nonrepudiation && (cert_usage & SC_CERT_USAGE_NONREPUDIATION)) {
+		pub_usage |= SC_PKCS15_PRKEY_USAGE_NONREPUDIATION;
+		pr_usage |= SC_PKCS15_PRKEY_USAGE_NONREPUDIATION;
+	}
+	/* filter usages algorithm */
+	if (pub_usage_ptr) {
+		*pub_usage_ptr = pub_usage & alg_flags;
+	}
+	if (pr_usage_ptr) {
+		*pr_usage_ptr = pr_usage & alg_flags;
+	}
+	return SC_SUCCESS;
+}
+
+static int sc_pkcs15emu_cac_init(sc_pkcs15_card_t *p15card)
+{
+	static const pindata pins[] = {
+		{ "1", NULL, "", 0x00,
+		  SC_PKCS15_PIN_TYPE_ASCII_NUMERIC,
+		  8, 4, 8, 
+		  SC_PKCS15_PIN_FLAG_NEEDS_PADDING |
+		  SC_PKCS15_PIN_FLAG_INITIALIZED ,
+		  -1, 0xFF,
+		  SC_PKCS15_CO_FLAG_PRIVATE },
+		{ NULL, NULL, NULL, 0, 0, 0, 0, 0, 0, 0, 0, 0}
+	};
+	/* oid for key usage */
+	static const u8 usage_type[] = { 0x55, 0x1d, 0x0f };
+	unsigned long long usage;
+
+
+	/*
+	 * The size of the key or the algid is not really known
+	 * but can be derived from the certificates. 
+	 * the cert, pubkey and privkey are a set. 
+	 * Key usages bits taken from certificate key usage extension.
+	 */
+
+	int    r, i;
+	sc_card_t *card = p15card->card;
+	sc_serial_number_t serial;
+	char buf[SC_MAX_SERIALNR * 2 + 1];
+	int count;
+	char *token_name = NULL;
+
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	memset(&serial, 0, sizeof(serial));
+
+	/* could read this off card if needed */
+
+	p15card->tokeninfo->label = strdup(cac_get_name(card->type)); 
+	p15card->tokeninfo->manufacturer_id = strdup(MANU_ID);
+	p15card->opts.use_file_cache = 1;
```

Comment:
It looks like that. I removed this line.

I am wondering, if there should not be some manual page for the `opensc.conf`. Its syntax is explained in the installed `/etc/opensc-x86_64.conf`, but having an formal overview would be plus for new users.

FYI, also `pkcs15-actalis.c` is setting this option for some reason.